Remove Uni await for OIDC code access token validation and user info calls

sberyozkin opened the pull request:
Fixes #13249.

This PR follows the earlier PR improving OIDC Uni related code:
- code access token verification and user info calls are totally async now; 
- user-info calls are always blocking hence they will run by the blocking executor
- code access token as well as a bearer access token will also be run by the blocking executor if the token is opaque - right now the introspection requests initiated internally by Vert.x would run on the IO thread

Changes look extensive but they are really not, it is just related to me having to collapse some code into reusable functions to minimize the duplication.

Also, `BlockingOperationControl` code branches will be gone as soon as I make sure that refreshing the tokens does not require a blocking call, but until then I have to check it to avoid re-dispatching to the executor (as explained by Stuart).

We already have the tests for all the affected combinations (code access token is a source of roles, opaque tokens, user-info)   

My Uni confidence is high now thanks to @cescoffier so this PR should be good to go once it will go green :-)

CC @cescoffier @gsmet

sberyozkin committed:
Remove Uni await for OIDC code access token validation and user info calls

gsmet commented:
@cescoffier could you have a close look to this one? We need to get this right.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -35,6 +36,11 @@
 public class OidcIdentityProvider implements IdentityProvider<TokenAuthenticationRequest> {
+
+    @SuppressWarnings("deprecation")
+    private static final Uni<AccessToken> NULL_CODE_ACCESS_TOKEN_UNI = Uni.createFrom().nullItem();
```
Why removing the deprecation?

sberyozkin replied:
@cescoffier What do you mean ? `AccessToken` is deprecated in Vert.x 3.8.x hence I add to avoid the IDE warnings...

cescoffier replied:
ah ok, I was referring to the mutiny part.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+            TenantConfigContext resolvedContext, AccessToken codeAccessToken) {
+
+        if (codeAccessToken != null) {
+            vertxContext.put("code_flow_access_token_result", codeAccessToken);
```
should be a constant.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                    @Override
+                    public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                        createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
```
even if blocking is allowed, it may be not called on the same thread. This method is called on the subscription thread, not the assembly thread.

sberyozkin replied:
@cescoffier I'm just trying to follow the same pattern as Stuart does [here](https://github.com/quarkusio/quarkus/blob/master/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L54). I can remove these `BlockingOperationContext` checks ?

sberyozkin replied:
And yes, it will be gone once that refresh token check becomes Uni async friendy, I'm actually nearly finished in another branch

cescoffier replied:
Except that in Stuart's code, the check is done at subscription time. You are doing it at assembly time and hope for the best are subscription:

```java
// This is called at assembly time
Uni.createFrom().emitter(e -> {
   // This is called at subscription time
});
```

The thread used at assembly and subscription time may be different. So if assembly time can block, it does not means you can block at subscription time (or the opposite).

sberyozkin replied:
Hey @cescoffier I'm on it, I'm learning something new here and at this rate I may qualify as a junior Mutiny bug fixer and start helping @jponge :-)

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
-                                    return;
-                                }
+                            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                                tenantResolver.getBlockingExecutor().execute(new Runnable() {
```
Any reason to not use Infrastructure.getDEfaultWorkerPool()?

sberyozkin replied:
@cescoffier It is the same executor which Stuart uses [here](https://github.com/quarkusio/quarkus/blob/master/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusIdentityProviderManagerImpl.java#L54) so I'm just trying to keep it consistent. Perhaps you'd like to open an issue for Stuart (or I can do it) to replace this executor with `Infrastructure.getDEfaultWorkerPool()` ?

cescoffier replied:
it's actually the same pool.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+            return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                @Override
+                public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                    createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
```
if I'm not mistaken this may be blocking, here you may be on the event loop.

sberyozkin replied:
@cescoffier please see [this comment](https://github.com/quarkusio/quarkus/pull/13257/files#r522163580) below. Unless it is an opaque token it should be fine to have on the IO thread, because it is a JWT token which is verified by a locally cached JWK keys

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+            TokenAuthenticationRequest request, TenantConfigContext resolvedContext, final JsonObject userInfo) {
+        resolvedContext.auth.decodeToken(request.getToken().getToken(),
+                new Handler<AsyncResult<AccessToken>>() {
+                    @Override
```
on which thread is this called? Are you 100% sure it cannot be the event loop?

cescoffier replied:
where is the blocking call you mentioned earlier?

sberyozkin replied:
@cescoffier Yeah; so if it is an `IdToken` which is a case with the code flow - then it is `JWT` => hence in 99% cases it is a local verification; if it is a bearer token (the one coming with HTTP `Authorization`) and it is JWT - the same situation. So it is OK to run it on the IO thread.

But if either token is opaque (binary) then it must be a remote introspection call done internally by `Vert.x` as part of this `decodeToken` call - hence in such cases it is a blocking call.

> where is the blocking call you mentioned earlier?

Right now, if OIDC is configured to refresh the token then it will run in the blocking executor starting from [here](https://github.com/quarkusio/quarkus/blob/master/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java#L62), before the decision to verify the token is made, hence this blocking operation control check

cescoffier replied:
Vert.x is doing a blocking call and block the event loop? That would be a first. @pmlopes, isn't this hidden?

cescoffier replied:
I guess that for you, as a user, you don't see the Vert.x gymnastics.

pmlopes replied:
I've no knowledge on the mentioned code, but I can assure that the calls from vert.x auth to the idp server are asynchronous. There's no `executeBlocking` code on vertx-auth-oauth2`

sberyozkin replied:
HI @cescoffier @pmlopes I think there may be some confusion now, I did not say `Vert.x` does a blocking call, I meant that since `introspection` calls are remote it is better for Quarkus OIDC to run them on the blocking executor, this is why we see so many blocked cases, as in case of the issue this PR is trying to fix, etc

cescoffier replied:
@sberyozkin That is a bit confusing. Why "since introspection calls are remote it is better for Quarkus OIDC to run them on the blocking executor"?  Where are these calls? In Vert.x (and so non-blocking) or in some other code?

cescoffier replied:
Is it just _easier_ for you to handle the validation? But in this case, how do you make the call (as you may mess up with the event loop used to execute the call and go back to blocking indefinitely).

sberyozkin replied:
@cescoffier the introspection calls are in Vert.x OAuth2, they are non-blocking (i.e async) but the introspection call is expensive, so we use the Executor for all such remote calls (introspection - starting from this PR, refresh tokens, OIDC connection resolution); IMHO it is OK, perhaps we can revisit it as part of a separate issue, but I believe it is not a bad idea...

sberyozkin replied:
@cescoffier the previous PR where you helped with splitting the tenant context resolution was related to a remote tenant connection resolution which was run on IO thread in one code branch...even though in Vert.x it is an async call

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+                                        resolvedContext.oidcConfig,
+                                        tokenJson, rolesJson, userInfo);
+                                if (tokenAutoRefreshPrepared(tokenJson, vertxContext, resolvedContext.oidcConfig)) {
+                                    throw new TokenAutoRefreshException(securityIdentity);
```
better failing and not throwing.

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -224,23 +286,62 @@ private static JsonObject getRolesJson(RoutingContext vertxContext, TenantConfig
+                            new Consumer<UniEmitter<? super AccessToken>>() {
+                                @Override
+                                public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                                    verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
```
again, may be a different thread. assembly and subscription threads different might be different.

sberyozkin replied:
@cescoffier please see the question above about this `BlockingOperationControl` check

cescoffier replied:
I don't believe it is.

sberyozkin replied:
@cescoffier This code just tries to run the remote operations in the executor, what is not correct about it ?

sberyozkin replied:
@cescoffier never mind, let me think about your `assembly vs subscription clarification` above

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -224,23 +286,62 @@ private static JsonObject getRolesJson(RoutingContext vertxContext, TenantConfig
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super AccessToken>>() {
+                    @Override
+                    public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                        verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
```
I don't get why sometimes it must be wrapped on an executor and sometimes not.

sberyozkin replied:
@cescoffier Please see above, [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522163580) and [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522168791) comments

cescoffier commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java:
```diff
@@ -266,33 +367,57 @@ public void handle(AsyncResult<AccessToken> event) {
+    private Uni<JsonObject> getUserInfoUni(RoutingContext vertxContext, TokenAuthenticationRequest request,
             TenantConfigContext resolvedContext) {
+        if (resolvedContext.oidcConfig.authentication.isUserInfoRequired()) {
+            if (BlockingOperationControl.isBlockingAllowed()) {
```
Again, same issue.

sberyozkin replied:
@cescoffier sure, hope the above comments, [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522163580) and [this](https://github.com/quarkusio/quarkus/pull/13257/files#r522168791) ones, will clarify it :-)

gsmet commented:
There seem to be a lot of questions and things to handle in this change.

Should we rather revert the original change for 1.10?

It doesn't look like an easy/obvious fix.

cescoffier commented:
@gsmet we can't. It was broken before. Now we enforce it, as it could literally kill the server at runtime (by blocking an event loop forever).

sberyozkin commented:
@gsmet It is OK, the original change was good (mostly, see below - but again not related to the current issue) and not related to this issue; the changes are simple once Clement clarified
@cescoffier I've updated the code, resolved those comments which I think I've addressed - reopen if needed please. 
This wrong pattern for checking the blocking context came from my earlier PR (see the updated `OidcRecorder` in this PR) - I wanted to write a good blocking call the way Stuart does it but missed a small detail :-)

cescoffier reviewed (approved):
The new changeset looks much better.

sberyozkin commented:
@cescoffier Thanks for your help Clement :-)

sberyozkin commented:
@gsmet hi Guillaume I'm going to merge this one tomorrow morning, Clement's approval is sufficient but I'll wait just in case Stuart or Pedro may want to add something

sberyozkin commented:
@jponge thanks
